kv/kvserver: TestTenantCtx failed

kv/kvserver.TestTenantCtx [failed](https://mesolite.cluster.engflow.com/invocations/default/6d58b26e-ceda-4d7f-a381-121fc584992c?testReportRun=1&testReportShard=45&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZzZXJ2ZXI6a3ZzZXJ2ZXJfdGVzdA==) on release-25.4.2-rc @ [ff51b7da52b4298ebd4282adccbb5f72f87ffb23](https://github.com/cockroachdb/cockroach/commits/ff51b7da52b4298ebd4282adccbb5f72f87ffb23):

Fatal error:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0xa345a6]

```
Stack: 

```
goroutine 2233501 [running]:
database/sql.(*Tx).grabConn(0x0, {0x8d4f6a8, 0xde5bbc0})
	GOROOT/src/database/sql/sql.go:2246 +0x46
database/sql.(*Tx).ExecContext(0x0, {0x8d4f6a8, 0xde5bbc0}, {0x7d12525, 0x1c}, {0xc005a57fb8, 0x1, 0x1})
	GOROOT/src/database/sql/sql.go:2504 +0x4b
database/sql.(*Tx).Exec(...)
	GOROOT/src/database/sql/sql.go:2517
github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.TestTenantCtx.func1.2()
	pkg/kv/kvserver/client_tenant_test.go:491 +0xd8
created by github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.TestTenantCtx.func1 in goroutine 1869531
	pkg/kv/kvserver/client_tenant_test.go:487 +0x599
```
<details><summary>Log preceding fatal error</summary>
<p>

```
        	            	  | 	pkg/rpc/auth.go:97
        	            	  | github.com/cockroachdb/cockroach/pkg/rpc.makeInternalClientAdapter.chainUnaryServerInterceptors.bindUnaryServerInterceptorToHandler.func4
        	            	  | 	pkg/rpc/context.go:873
        	            	  | github.com/cockroachdb/cockroach/pkg/rpc.gatewayRequestRecoveryInterceptor
        	            	  | 	pkg/rpc/metrics.go:514
        	            	  | github.com/cockroachdb/cockroach/pkg/rpc.makeInternalClientAdapter.chainUnaryServerInterceptors.bindUnaryServerInterceptorToHandler.func4
        	            	  | 	pkg/rpc/context.go:873
        	            	  | github.com/cockroachdb/cockroach/pkg/server.newGRPCServer.NewRequestMetricsInterceptor.func4
        	            	  | 	pkg/rpc/metrics.go:467
        	            	  | github.com/cockroachdb/cockroach/pkg/rpc.makeInternalClientAdapter.chainUnaryServerInterceptors.bindUnaryServerInterceptorToHandler.func4
        	            	  | 	pkg/rpc/context.go:873
        	            	  | github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func1.1
        	            	  | 	pkg/rpc/context.go:139
        	            	  | github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunTaskWithErr
        	            	  | 	pkg/util/stop/stopper.go:351
        	            	  | github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func1
        	            	  | 	pkg/rpc/context.go:137
        	            	  | github.com/cockroachdb/cockroach/pkg/rpc.makeInternalClientAdapter.chainUnaryServerInterceptors.bindUnaryServerInterceptorToHandler.func4
        	            	  | 	pkg/rpc/context.go:873
        	            	  | github.com/cockroachdb/cockroach/pkg/rpc.makeInternalClientAdapter.func2
        	            	  | 	pkg/rpc/context.go:772
        	            	  | github.com/cockroachdb/cockroach/pkg/rpc.NewContext.ClientInterceptor.func9
        	            	  | 	pkg/util/tracing/grpcinterceptor/grpc_interceptor.go:185
        	            	  | github.com/cockroachdb/cockroach/pkg/rpc.getChainUnaryInvoker.func1
        	            	  | 	pkg/rpc/context.go:957
        	            	  | github.com/cockroachdb/cockroach/pkg/rpc.makeInternalClientAdapter.func3
        	            	  | 	pkg/rpc/context.go:842
        	            	  | github.com/cockroachdb/cockroach/pkg/rpc.internalClientAdapter.Batch
        	            	  | 	pkg/rpc/context.go:965
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*grpcTransport).sendBatch
        	            	  | 	pkg/kv/kvclient/kvcoord/transport.go:211
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*grpcTransport).SendNext
        	            	  | 	pkg/kv/kvclient/kvcoord/transport.go:189
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).sendToReplicas
        	            	  | 	pkg/kv/kvclient/kvcoord/dist_sender.go:2768
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).sendPartialBatch
        	            	  | 	pkg/kv/kvclient/kvcoord/dist_sender.go:2227
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).divideAndSendBatchToRanges
        	            	  | 	pkg/kv/kvclient/kvcoord/dist_sender.go:1775
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).Send
        	            	  | 	pkg/kv/kvclient/kvcoord/dist_sender.go:1280
        	            	Wraps: (2) expected Scan to run as the expected tenant ({10}), but it isn't. tenant request: false, tenantID: {0}
        	            	Error types: (1) *withstack.withStack (2) *errutil.leafError
        	Test:       	TestTenantCtx/shared-process_tenant=true
    client_tenant_test.go:490: 
        	Error Trace:	pkg/kv/kvserver/client_tenant_test.go:490
        	            				src/runtime/asm_amd64.s:1700
        	Error:      	Received unexpected error:
        	            	dial tcp 127.0.0.1:46429: connect: connection reset by peer
        	Test:       	TestTenantCtx/shared-process_tenant=true
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>deadlock=true</code>
 - <code>run=1</code>
 - <code>shard=45</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv-triage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTenantCtx.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-57338

--- a/pkg/kv/kvserver/client_tenant_test.go
+++ b/pkg/kv/kvserver/client_tenant_test.go
@@ -417,8 +417,14 @@ func TestTenantCtx(t *testing.T) {
 						// context looks as expected, and signal their channels.
 
 						tenID, isTenantRequest := roachpb.ClientTenantFromContext(ctx)
-						keyRecognized := strings.Contains(ba.Requests[0].GetInner().Header().Key.String(), magicKey)
+						key := ba.Requests[0].GetInner().Header().Key
+						keyRecognized := strings.Contains(key.String(), magicKey)
 						if !keyRecognized {
+							return nil
+						}
+
+						// Skip meta2 range lookups as they don't carry tenant context.
+						if bytes.HasPrefix(key, keys.Meta2Prefix) {
 							return nil
 						}
 
@@ -430,6 +436,9 @@ func TestTenantCtx(t *testing.T) {
 						if isSinglePushTxn := ba.IsSingleRequest() && ba.Requests[0].GetInner().Method() == kvpb.PushTxn; isSinglePushTxn {
 							pushReq = ba.Requests[0].GetInner().(*kvpb.PushTxnRequest)
 						}
+
+						t.Logf("received request with key: %s, isTenantRequest: %t, tenID: %d",
+							key.String(), isTenantRequest, tenID)
 
 						switch {
 						case scanReq != nil:

